addons/survey/controllers/main.py

```diff
@@ -732,35 +727,130 @@ def _get_user_input_domain(self, survey, line_filter_domain, **post):
 
-        user_input_domain = self._get_user_input_domain(survey, line_filter_domain, **post)
-        user_input_lines = request.env['survey.user_input'].sudo().search(user_input_domain).mapped('user_input_line_ids')
+        answer_cols_to_rows_mapping, answer_user_input_line_ids = self._get_filters_from_post(post)
+
+        # Matrix, Multiple choice, Simple choice filters
+        if answer_cols_to_rows_mapping:
+            ans_ids, row_ids = [], []
+            for filter_ans_id, filter_row_ids in answer_cols_to_rows_mapping.items():
+                ans_ids.append(filter_ans_id)
+                row_ids += filter_row_ids
+
+            answers = request.env['survey.question.answer'].browse(list(filter(bool, ans_ids+row_ids)))  # exclude '0'
+            # For performance, caches all useful columns of survey_question_answer in first query
+            answers_only = answers.filtered(lambda a: not a.matrix_question_id)   # not row_ids
+
+            for answer in answers_only:
+                for row_id in answer_cols_to_rows_mapping[answer.id]:
+                    filters_domains += self._prepare_matching_answer_domain(answer.question_type, answer, row_id=row_id)
+                    search_filters.append(self._prepare_search_filter('A', answer, row_id=row_id))
+
+        # Char_box, Text_box, Numerical_box, Date, Datetime filters
+        if answer_user_input_line_ids:
+            user_input_line_ids = request.env['survey.user_input.line'].browse(answer_user_input_line_ids)
+            for input_line_id in user_input_line_ids:
+                value, operator = self._get_value_with_operator(input_line_id)
+                filters_domains += self._prepare_matching_answer_domain(input_line_id.answer_type, input_line_id, operator=operator, value=value)
+                search_filters.append(self._prepare_search_filter('L', input_line_id, value=value))
+
+        user_input_domain = self._get_results_page_user_input_domain(survey, **post)
+        user_input_lines = self._get_matching_user_input_lines(filters_domains, user_input_domain)
 
         return user_input_lines, search_filters
+
+    def _get_matching_user_input_lines(self, filters_domains, user_input_domain):
+        """ Retrieve the matching user input lines depending on the survey base domain and the filters domains.
+
+        :return recordset: all matching user input lines
+        """
+        # Add filters domain to the base domain
+        if filters_domains:
+            all_required_lines_domains = [
+                [('user_input_line_ids', 'in', request.env['survey.user_input.line'].sudo()._search(filters_domain))]
+                for filters_domain in filters_domains
+            ]
+            user_input_domain = expression.AND([user_input_domain, *all_required_lines_domains])
+
+        # Get the matching user input lines
+        user_inputs_query = request.env['survey.user_input'].sudo()._search(user_input_domain)
```

Comment:
Could we do it the other way around?

Add a _search on the `('user_input_id', 'in', request.env['survey.user_input'].sudo()._search(user_input_domain))`.
Then mix it with "all_required_lines_domains".

Then search the lines we want.
Seems more simple but maybe I missed something.

Reply:
No this is correct actually.

It's just a bit confusing because we make a domain on user_input_lines, to retrieve user_inputs, to then retrieve user_input_lines. So it makes you wonder "why not directly retrieving user_input_lines" ?
But it would not work because we want ALL lines tied to the user_inputs matching ANY of the retrieved lines.

Maybe elaborate a bit more on that in the method docstring? :slightly_smiling_face: